src/metrics.py

```diff
@@ -0,0 +1,380 @@
+    data_dict = {
+        "k": [],
+        "q": [],
+        "v": [],
+        "in_proj": [],
+        "ffn_in": [],
+        "ffn_out": [],
+    }
+    for i, layer in enumerate(model.blocks):
+        k, q, v, in_proj, ffn_in, ffn_out = get_layer_metrics_transformer(layer, nheads)
+        data_dict["k"].append({i: k})
+        data_dict["q"].append({i: q})
+        data_dict["v"].append({i: v})
+        data_dict["in_proj"].append({i: in_proj})
+        data_dict["ffn_in"].append({i: ffn_in})
+        data_dict["ffn_out"].append({i: ffn_out})
+
+    weights = torch.cat(
+        [concatenate_matrices(unpack_weights_transformer(layer)) for layer in model.blocks]
+    )
+    biases = torch.cat(
+        [concatenate_matrices(unpack_biases_transformer(layer)) for layer in model.blocks]
+    )
+
+    data_dict["w_all"] = get_distribution_stats(weights)
+    data_dict["b_all"] = get_distribution_stats(biases)
+
+    return data_dict
+
+
+def get_distribution_stats(X):
+    mean = torch.mean(X).item()
+    var = torch.var(X).item()
+    median = torch.median(X).item()
+    
+    return {"mean": mean, "var": var, "median": median}
+
+
+@torch.no_grad()
+def get_metrics_mlp(model):
+    data_dict = {
+        "w": [],
+    }
+
+    weights = []
+    biases = []
+
+    for i, layer in enumerate(model.layers):
+        if isinstance(layer, nn.Linear):
+            w = get_layer_metrics_mlp(layer)
+            data_dict["w"].append({i: w})
+            weights.append(layer.weight.flatten())
+            biases.append(layer.bias.flatten())
+
+    weights = torch.cat(weights)
+    biases = torch.cat(biases)
+
+    data_dict["w_all"] = get_distribution_stats(weights)
+    data_dict["b_all"] = get_distribution_stats(biases)
+
+    return data_dict
+
+
+@torch.no_grad()
+def get_metrics_hf_transformer(model):
+    data_dict = {
+        "k": [],
+        "q": [],
+        "v": [],
+        "in_proj": [],
+        "ffn_in": [],
+        "ffn_out": [],
+    }
+
+    for i, layer in enumerate(model.bert.encoder.layer):
+        k, q, v, in_proj, ffn_in, ffn_out = get_layer_metrics_hf_transformer(layer)
+        data_dict["k"].append({i: k})
+        data_dict["q"].append({i: q})
+        data_dict["v"].append({i: v})
+        data_dict["in_proj"].append({i: in_proj})
+        data_dict["ffn_in"].append({i: ffn_in})
+        data_dict["ffn_out"].append({i: ffn_out})
+
+    weights = torch.cat(
+        [
+            concatenate_matrices(unpack_weights_hf_transformer(layer))
+            for layer in model.bert.encoder.layer
+        ]
+    )
+    biases = torch.cat(
+        [
+            concatenate_matrices(unpack_biases_hf_transformer(layer))
+            for layer in model.bert.encoder.layer
+        ]
+    )
+
+    data_dict["w_all"] = get_distribution_stats(weights)
+    data_dict["b_all"] = get_distribution_stats(biases)
+
+    return data_dict
+
+
+def induction_head_loss(losses):
+    return (losses[..., 500] - losses[..., 50]).mean()
+
```

Comment:
Can you add a comment explaining why you're indexing at indices 500 and 50 in particular?